### PR TITLE
fix(controller): check storedTemplates in GetNodeTemplate before live CWT lookup

### LINKED ISSUE #15655: Cross-CWT templateRef node spuriously marked Error when CWT is modified after pod creation
### Pre-requisites

- [x] I have double-checked my configuration
- [x] I have tested with the `:latest` image tag (i.e. `quay.io/argoproj/workflow-controller:latest`) and can confirm the issue still exists on `:latest`. If not, I have explained why, **in detail**, in my description below.
- [x] I have searched existing issues and could not find a match for this bug
- [ ] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/main/docs/CONTRIBUTING.md))

### What happened? What did you expect to happen?

**What happened:**

When a workflow uses a cross-CWT `templateRef` (a DAG task in CWT-A referencing a template in CWT-B), and the referenced template is removed from CWT-B while the pod is already running, the workflow controller marks the pod node as `Error` with:

```
template <template-name> not found in workflow template <cwt-name>
```

This is a **spurious error** — the pod itself ran and completed normally. The error is set after the pod finishes, during `assessNodeStatus`.

**Root cause:**

`assessNodeStatus` calls `GetNodeTemplate`, which for nodes with `templateRef` calls `GetTemplateFromRef` directly against the CWT informer cache — bypassing `wf.Status.StoredTemplates`. The template is correctly snapshotted in `storedTemplates` at workflow start (because `resolveTemplateImpl` / `executeTemplate` checks `storedTemplates` first), but `GetNodeTemplate` does not perform this check.

The two paths behave differently:

- **Execution path** (`executeTemplate` → `resolveTemplateImpl`): checks `storedTemplates` first ✅
- **Pod observation path** (`assessNodeStatus` → `GetNodeTemplate` → `GetTemplateFromRef`): skips `storedTemplates`, goes straight to the CWT informer cache ❌

`GetNodeTemplate` for `templateName` nodes is unaffected — it calls `woc.wf.GetTemplateByName()` which searches `storedWorkflowSpec` (snapshotted). Only `templateRef` nodes are vulnerable.

**Expected:**

The node should reflect the actual pod outcome (Succeeded/Failed). The controller should use the snapshotted template from `storedTemplates` rather than failing when the live CWT no longer contains the template.

**Proposed fix** (`workflow/controller/operator.go`):

```go
func (woc *wfOperationCtx) GetNodeTemplate(node *wfv1.NodeStatus) (*wfv1.Template, error) {
    if node.TemplateRef != nil {
        // Check storedTemplates first (mirrors resolveTemplateImpl)
        scope, resourceName := node.GetTemplateScope()
        if tmpl := woc.wf.GetStoredTemplate(scope, resourceName, node); tmpl != nil {
            return tmpl, nil
        }
        // Fall back to live CWT informer
        tmplCtx, err := woc.createTemplateContext(node.GetTemplateScope())
        if err != nil {
            return nil, err
        }
        return tmplCtx.GetTemplateFromRef(node.TemplateRef)
    }
    return woc.wf.GetTemplateByName(node.TemplateName), nil
}
```

**Related issues:** #8956 documents the same call chain (`assessNodeStatus` → `GetNodeTemplate` → `GetTemplateFromRef`) panicking when the CWT informer cache is unsynced after HA leader election (different root cause, closed).

### Version(s)

v4.0.1

### Paste a minimal workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflow that uses private images.

**Reproduction steps:**
1. Apply the two ClusterWorkflowTemplates and the Workflow below.
2. Wait until the pod for `hello-task` is in **Running** state.
3. Remove the `say-hello` template from `cwt-helper`: `kubectl edit clusterworkflowtemplate cwt-helper` (delete the template entry from `spec.templates`).
4. Wait for the pod to complete.

**Result:** The `hello-task` node is marked `Error` — "template say-hello not found in workflow template cwt-helper" — even though the pod completed successfully.

```yaml
# Step 1: CWT containing the referenced template
apiVersion: argoproj.io/v1alpha1
kind: ClusterWorkflowTemplate
metadata:
  name: cwt-helper
spec:
  templates:
    - name: say-hello
      script:
        image: alpine:3.18
        command: [sh]
        source: |
          echo "hello"
          sleep 30
---
# Step 2: CWT with a DAG that cross-references cwt-helper
apiVersion: argoproj.io/v1alpha1
kind: ClusterWorkflowTemplate
metadata:
  name: cwt-main
spec:
  templates:
    - name: main
      dag:
        tasks:
          - name: hello-task
            templateRef:
              name: cwt-helper
              template: say-hello
              clusterScope: true
---
# Step 3: Workflow
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: cwt-templateref-bug-
spec:
  entrypoint: main
  workflowTemplateRef:
    name: cwt-main
    clusterScope: true
```

### Logs from the workflow controller

```text
kubectl logs -n argo deploy/workflow-controller | grep ${workflow}
```

Expected log line when the bug triggers:

```
template say-hello not found in workflow template cwt-helper
```

### Logs from in your workflow's wait container

```text
kubectl logs -n argo -c wait -l workflows.argoproj.io/workflow=${workflow},workflow.argoproj.io/phase!=Succeeded
```

### PR DESCRIPTION
**Pull request by nancysangani**
Fixes #15655

When a cross-CWT templateRef node's template is removed from the CWT
after the pod is already running, assessNodeStatus was spuriously marking
the node Error with "template not found" — even though the pod completed
successfully.

The fix makes GetNodeTemplate check wf.Status.StoredTemplates first
(mirroring what resolveTemplateImpl already does during execution) before
falling back to the live CWT informer cache. The fallback is kept for
edge cases like controller restarts.

Added a unit test that reproduces the bug by leaving the live CWT cache
empty while the template exists in storedTemplates.

### CODE REVIEW

File: docs/quick-start.md [hunk 1]
```diff
@@ -31,7 +31,7 @@ Then, copy the commands below to apply the quick-start manifest:
 
 ```bash
 kubectl create namespace argo
-kubectl apply -n argo -f "https://github.com/argoproj/argo-workflows/releases/download/${ARGO_WORKFLOWS_VERSION}/quick-start-minimal.yaml"
+kubectl apply --server-side -n argo -f "https://github.com/argoproj/argo-workflows/releases/download/${ARGO_WORKFLOWS_VERSION}/quick-start-minimal.yaml"
```
[the remark this answers is not in the document]
Comment (thevilledev) [reply docs/quick-start.md:34]: Separate issue from this PR.

File: workflow/controller/operator_test.go [hunk 2]
```diff
@@ -1931,6 +1931,57 @@ func TestAssessNodeStatus(t *testing.T) {
...
+
+	ctx := logging.TestContext(t.Context())
+	cancel, controller := newController(ctx, wf)
+	defer cancel()
+	woc := newWorkflowOperationCtx(ctx, wf, controller)
+
+	tmpl, err := woc.GetNodeTemplate(ctx, node)
+
+	require.NoError(t, err, "GetNodeTemplate must not error when template is in storedTemplates")
+	require.NotNil(t, tmpl, "GetNodeTemplate must return the stored template")
```
Comment (thevilledev) [inline workflow/controller/operator_test.go]: Consider adding a case where `StoredTemplates` is empty but the CWT exists in the informer cache, to verify the live lookup still works.

**Review by thevilledev — COMMENTED** [review]
Minor test related comment.

File: docs/quick-start.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (nancysangani) [reply docs/quick-start.md:34]: Done, reverted the unrelated docs change. The PR now only contains changes to workflow/controller/operator.go and workflow/controller/operator_test.go.

File: workflow/controller/operator_test.go (hunk 2 above)
Comment (nancysangani) [reply workflow/controller/operator_test.go]: Added a second subtest that verifies the live CWT informer fallback works when StoredTemplates is empty. Both cases pass locally.

**Comment by nancysangani** [comment]
@thevilledev Please review it when you get a chance, I have implemented the suggested changes. Thanks!

**Review by thevilledev — APPROVED** [review]
Thank you, LGTM. Pending for maintainer feedback

**Comment by deynekas** [comment]
What is the status of this please? 
In our case we experienced following behaviour caused by same issue:
  1. Multiple workflows were running, executing steps that reference `ClusterWorkflowTemplate/workflow-v1`
  2. We deployed an update to the CWT `workflow-v1` that added a new output parameter: `jenkinsBuildNumber`
  3. Workflows submitted BEFORE the change hung indefinitely even though `workflow-v1` completed successfully

From my POV This bug fundamentally breaks the immutability guarantee that `storedTemplates` was designed to provide. Template changes should NOT affect in-flight workflows, but currently they do because:

  - ✅ **Execution path**: Correctly uses `storedTemplates` (immutable)
  - ❌  **Assessment path**: Queries live CWT (mutable) ← **This PR fixes this**
  
/cc @Joibel
/cc @thevilledev

**Comment by Kolajik** [comment]
👋 Hello, any update on this issue?

We're experiencing the same issue @deynekas reported. It would help us tremendously with making changes to any `WorkflowTemplate` such that it doesn't affect running workflows. Thanks!

**Comment by Joibel** [comment]
#15659 is the same fix, but with better tests. Closing this, preferring that one.
